databroker/core.py

```diff
@@ -673,3 +735,97 @@ def _external_keys(descriptor, _cache=boltons.cacheutils.LRU(max_size=500)):
         ek = {k: v.get('external', None) for k, v in data_keys.items()}
         _cache[descriptor['uid']] = ek
     return ek
+
+
+class DocBuffer:
+    '''Buffer a (name, document) sequence into parts
+
+    '''
+    def __init__(self, doc_gen, denormalize=False):
+
+        class InnerDict(dict):
+            def __getitem__(inner_self, key):
+                while key not in inner_self:
+                    try:
+                        self._get_next()
+                    except StopIteration:
+                        raise Exception("this stream does not contain a "
+                                        "descriptor with uid {}".format(key))
+                return super().__getitem__(key)
+
+        self.denormalize = denormalize
+        self.gen = doc_gen
+        self._start = None
+        self._stop = None
+        self.descriptors = InnerDict()
+        self._events = deque()
+
+    @property
+    def start(self):
+        while self._start is None:
+            try:
+                self._get_next()
+            except StopIteration:
+                raise Exception("stream does not contain a start?!")
+
+        return self._start
+
+    @property
+    def stop(self):
+        while self._stop is None:
+            try:
+                self._get_next()
+            except StopIteration:
+                raise Exception("stream does not contain a start")
```

Comment:
Ha. You said it so I didn't have to. I'm completely sold on your argument to move toward inheriting from package-specific base exceptions, btw. Not sure if this PR is the right place to start that crusade, but we should do it soon.